### PR TITLE
[demangle] Add a new tool called swift-demangle-yamldump.

### PR DESCRIPTION
**Pull request by gottesmm**
This is a separate tool so we can use all of LLVM without worrying about
creating a dependency from the demangler on LLVM's libsupport.

I am going to use it to fix cmpcodesize for the new mangling by eliminating
cmpcodesize's usage of regex to classify symbols. Instead, we can just read in
the yaml version of the symbol trees for each demangled node and process that
instead.

rdar://41146023

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
Missed the test for this... oops.

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

File: tools/swift-demangle-yamldump/swift-demangle-yamldump.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+#include "llvm/Support/Signals.h"
+#include "llvm/Support/raw_ostream.h"
+
+// For std::rand, to work around a bug if main()'s first function call passes
+// argv[0].
+#if defined(__CYGWIN__)
+#include <cstdlib>
+#endif
+
+#include <iostream>
```
Comment (jrose-apple) [inline tools/swift-demangle-yamldump/swift-demangle-yamldump.cpp]: Nitpick: you're not using this, right? And it has a static constructor.

File: tools/swift-demangle-yamldump/swift-demangle-yamldump.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    hadLeadingUnderscore = true;
+    name = name.substr(1);
+  }
+  swift::Demangle::NodePointer pointer = DCtx.demangleSymbolAsNode(name);
+  // We do not emit a message so that we end up dumping
+  llvm::outs() << getNodeTreeAsYAML(name, pointer);
+  DCtx.clear();
+}
+
+static int demangleSTDIN(const swift::Demangle::DemangleOptions &options) {
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply tools/swift-demangle-yamldump/swift-demangle-yamldump.cpp]: I agree with the c++filt (i.e. trying to match), but I disagree about not accepting input from stdin. It avoids problems with long command lines. Maybe instead the tool should abort if it fails to demangle.

[the remark this answers is not in the document]
Comment (jrose-apple) [reply tools/swift-demangle-yamldump/swift-demangle-yamldump.cpp]: \*shrug\* Okay. The main thing is that there's not really a reason to accept *arbitrary* input, which swift-demangle does have.

**Pull request merged**